### PR TITLE
docs - add db privileges, update db connection troubleshooting

### PR DESCRIPTION
**Pull request by nllho**
Started this to add a note to check the db role + privileges. Jeff's feedback got me thinking about adding a section on the necessary user/role/privileges to the db connection doc.

---

UPDATE: Lots of new info! Enough for a fresh doc :)

I focused on the idea of "using roles to bundle privileges based on use cases".

Didn't test the SQL -- only modified @likeshumidity's SQL examples very slightly to use different names (and sometimes execute the role vs. user creation in a different order). Luckily, in Postgres, a user and a role are pretty much interchangeable!

To do:
- [x] Add connective tissue to other docs.

- - -
This change is [Reviewable](https://reviewable.io/reviews/metabase/metabase/30205)

### CODE REVIEW

File: docs/troubleshooting-guide/db-connection.md [hunk 1]
```diff
@@ -34,6 +34,8 @@ If you don't have access to the Metabase Admin panel, you'll need to ask the per
 
     - If you're running Metabase Cloud, check that you've [whitelisted our IP addresses](https://www.metabase.com/cloud/docs/ip-addresses-to-whitelist).
 
+3. Make sure that Metabase is using a role with the correct privileges to connect to your data warehouse.
```
Comment (jeff-bruemmer) [inline docs/troubleshooting-guide/db-connection.md:37]:
```suggestion
3. Make sure that Metabase is connecting to your database using a database user account that has a role with the privileges to query and write to the tables that you want to query or write to. If the database user account doesn't have a role with those privileges, you'll either need to grant that account a role (or roles) with the necessary privileges, or have Metabase connect using a different database account.
```

Is this more precise? Just want to clarify that, like Metabase's accounts and groups, databases typically have have users and roles.

**Review by jeff-bruemmer — APPROVED** [review]

File: docs/databases/connecting.md [hunk 2]
```diff
@@ -45,6 +45,70 @@ For provider-specific connection details, like connecting to a PostgreSQL data w
 
 - [AWS's Relational Database Service (RDS)](./connections/aws-rds.md)
 
+## Database roles, users, and privileges
+
+For Metabase to connect, query, or write to your database, you must give Metabase a database user account with the correct database privileges.
+
+The easiest way to set this up in a Postgres database:
```
Comment (jeff-bruemmer) [inline docs/databases/connecting.md:52]: The easiest (but not recommended)?

File: docs/databases/connecting.md [hunk 3]
```diff
@@ -45,6 +45,70 @@ For provider-specific connection details, like connecting to a PostgreSQL data w
...
+> If you're setting up multi-tenant permissions for people who need SQL access to specific schemas or tables, you may want to create [multiple database users](https://www.metabase.com/learn/permissions/multi-tenant-permissions#option-2-granting-customers-native-sql-access-to-their-schema) for Metabase, such as `metabase_customer_1` and `metabase_customer_2`.
+
+### Database privileges
+
+Once you've set up your [database roles and users](#database-roles-and-users), you can assign database privileges to those roles. When you grant privileges to a role, all users with that role will get those privileges.
+
+At minimum, the `bi_tool` role should be able to connect to and query your database:
+
+```sql
+GRANT CONNECT ON DATABASE "your_database" TO bi_tool;
```
Comment (jeff-bruemmer) [inline docs/databases/connecting.md:94]: I like this second pass, but think we could go further. We should give people more complete code here, so people don't have to figure this out for themselves and can just copy/paste/modify.

In fact, we could/should recommend a cookie cutter set of SQL and roles to set up for Metabase. Like: tell them to create specific roles that have different privileges required for different Metabase features.

Something like this pseudocode:

```
-- Set up db user account that Metabase will use to connect to the database.
CREATE USER metabase;

-- Create roles with different privileges
--
-- Role with the minimum privileges for Metabase to work
CREATE ROLE connect_and_query;
GRANT CONNECT ON DATABASE "your_database" TO connect_and_query;
...

-- Role for Metabase actions and model caching to work
-- (Nat, I'm guessing on this here, we may need to split model caching and action privileges into separate roles)
CREATE ROLE actions_and_caching
GRANT CONNECT ...

-- Assign role(s) to `metabase` user account.
GRANT connect_and_query to metabase;
GRANT actions_and_caching to metabase;
```

And then, like you do below, also show how to do it for particular schema/tables.

**Review by jeff-bruemmer — CHANGES_REQUESTED** [review]
@nllho this is great so far.

File: docs/databases/connecting.md [hunk 4]
```diff
@@ -45,6 +45,70 @@ For provider-specific connection details, like connecting to a PostgreSQL data w
...
+GRANT USAGE ON SCHEMA "your_schema" TO bi_tool;
+GRANT SELECT ON "your_table" IN SCHEMA "your_schema" TO bi_tool;
+```
+
+If you plan on using [model caching](../data-modeling/models.md#model-caching) or [actions](../actions/introduction.md) in Metabase, you'll need to give the `metabase` user extra privileges to write to the table used in your model:
+
+```sql
+GRANT USAGE ON SCHEMA "your_schema" TO metabase;
+GRANT INSERT, UPDATE, DELETE ON "your_model's_table" IN SCHEMA "your_schema" TO metabase;
+```
```
[the remark this answers is not in the document]
Comment (likeshumidity) [reply docs/databases/connecting.md:110]: The below should work with two roles. One for just read only analytics access and the other for analytics with writeback (CREATE access necessary for model caching):

```sql
-- create user and password for Metabase
CREATE USER metabase WITH PASSWORD 'INSERT_PASSWORD';
-- create role for read-only analytics
CREATE ROLE analytics_read_only WITH LOGIN;
-- grant SELECT access to to all tables for role
GRANT SELECT ON ALL TABLES IN SCHEMA public TO analytics_read_only;
-- grant SELECT access to to specific tables for role
GRANT SELECT ON public.model_actions_test TO analytics_read_only;
-- grant role to Metabase user
GRANT analytics_read_only TO metabase;


-- Add the below for model caching
-- create role for analytics with writeback access to database for model caching
CREATE ROLE analytics_with_writeback WITH LOGIN;
-- grant CREATE access to database
GRANT CREATE ON DATABASE "testing-20230207a" TO analytics_with_writeback;
-- grant INSERT, UPDATE, DELETE access to all tables in schema (so it has access to newly created tables)
GRANT SELECT, INSERT, UPDATE, DELETE ON ALL TABLES IN SCHEMA public TO analytics_with_writeback;
-- grant role to Metabase user
GRANT analytics_with_writeback TO metabase;
```

File: docs/databases/users-roles-privileges.md [hunk 5]
```diff
@@ -0,0 +1,192 @@
...
+
+-- Option 3: Let users with the analytics role query anything in a specific TABLE.
+-- GRANT SELECT ON ALL TABLES IN SCHEMA "schema" TO analytics;
+```
+
+Remember that when you grant privileges to a role, all users with that role will get those privileges.
+
+## Minimum database privileges
+
+In order to view and query your tables in Metabase, you'll have to give Metabase's database user:
```
Comment (jeff-bruemmer) [inline docs/databases/users-roles-privileges.md:67]: Can we get the SQL here for the lazy (following the grant to role, then grant role to user pattern)?

File: docs/databases/users-roles-privileges.md [hunk 6]
```diff
@@ -0,0 +1,192 @@
...
+-- Create a database user named "metabase".
+CREATE USER metabase WITH PASSWORD "your_password";
+
+-- Give the user read and write privileges to anything in the database.
+GRANT ALL PRIVILEGES ON "database" TO metabase;
+```
+
+This option is great if you're connecting to a local database for development or testing.
+
+## Actions
```
Comment (jeff-bruemmer) [inline docs/databases/users-roles-privileges.md:96]:
```suggestion
## Privileges Metabase needs to run actions
```

File: docs/databases/users-roles-privileges.md [hunk 7]
```diff
@@ -0,0 +1,192 @@
...
+title: Database users, roles, and privileges
+---
+
+# Database users, roles, and privileges
+
+Metabase needs some [minimum database privileges](#minimum-database-privileges) to successfully connect to your database, as well as some additional database privileges for features like [actions](#actions) and [model caching](#model-caching).
+
+The structure of your database privileges should fit into your [permissions strategy](https://www.metabase.com/learn/permissions/strategy).
+
+## Organize database privileges with roles
```
Comment (jeff-bruemmer) [inline docs/databases/users-roles-privileges.md:11]: Think we need one more short section, as close to the top of the page as possible, that just gives people a set-and-forget SQL code block they can copy and paste to get moving if they don't want to bother with all this reading and fine-tuning. Something like: create a role that has/lists all the privileges Metabase would need, create a metabase user, grant role to user, done. Like the "Grant all privileges" section, but more explicit about the privileges granted, i.e., this should state every privilege Metabase would need, including `TEMPORARY`, etc. Might even want to use language like "set and forget", but then encourage/recommend the more nuanced approach.

The bonus side effect of frontloading this short-but-explicit version is that the list of grants will let experienced database admins see at a glance exactly what they need, and they can run with it and set up privileges however they want.

**Review by jeff-bruemmer — APPROVED** [review]
Nice revision/expansion on this, @nllho. Approving, but see my comments before merging. 

For the larger asks, you can either tact them onto this PR or create a new PR for me to review if you like, but this is already way ahead of our current resources on this topic.

**Pull request merged**